book/src/developers/designs/oauth2_refresh_tokens.md

```diff
@@ -0,0 +1,161 @@
+     │Access │              │    │Access │                         │               │Access │           
+     │   +   │              │    │   +   │                         │               │   +   │           
+     │Refresh│──3. Refresh──┘    │Refresh│                         │               │Refresh│           
+     │ IAT 1 │                   │ IAT 2 │─────6. Refresh──────────┘               │ IAT 1 │           
+     │       │                   │       │                                         │       │           
+     └───────┘                   └───────┘                                         └───────┘
+
+In this example, we can see that the replication of the session with NIB 1 happens to the second
+Kanidm server, but the replication of session with NIB 2 has not occurred yet. If the token that was
+later issued with IAT 2 was presented to the second server it would still be valid and able to
+refresh since IAT 2 is greater or equal to NIB 1. This would also prompt the session to advance to
+NIB 3 such that when replication begun again, the session with NIB 3 would take precedence over the
+former NIB 2 session.
+
+While this allows a short window where a former access token could be used on the second replica,
+this infrastructure being behind load balancers and outside of an attackers influence significantly
+hinders the ability to attack this for very little gain.
+
+## Attack Detection
+
+[draft oauth security topics section 4.14.2](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#section-4.14.2)
+specifically calls out that when refresh token re-use is detected then all tokens of the session
+should be canceled to cause a new authorisation code flow to be initiated.
+
+## Inactive Refresh Tokens
+
+Similar
+[draft oauth security topics section 4.14.2](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#section-4.14.2)
+also discusses that inactive tokens should be invalidated after a period of time. From the view of
+the refresh token this is performed by an internal exp field in the encrypted refresh token.
+
+From the servers side we will require a "not after" parameter that is updated on token activity.
+This will also require inactive session cleanup in the server which can be extended into the session
+consistency plugin that already exists.
+
+Since the act of refreshing a token is implied activity then we do not require other signaling
+mechanisms.
+
+# Questions
+
+Currently with authorisation code grants and sessions we issue these where the sessions are recorded
+in an async manner. For consistency I believe the same should be true here but is there a concern
+with the refresh being issued but a slight delay before it's recorded? I think given the nature of
+our future replication we already have to consider the async/eventual nature of things, so this
+doesn't impact that further, and may just cause client latency in the update process.
```

Comment:
That's something to measure when the replication's built I guess - it's the kind of thing you'd want a completion contract internally for these kinds of high-risk write transactions. 

If it's going to take more than x seconds to write a session out locally (ignoring replication delays), then just go into blocking mode because something's going wrong and it's better to hold off on auth attempts than allow race conditions to occur.

Reply:
Agreed, I think i'll just put this into the write txn path. Our writes are extremely fast anyway so I think it's not an issue.